src/openforms/formio/validation.py

```diff
@@ -0,0 +1,131 @@
+"""
+from typing import Any, Callable
+
+from django.core.exceptions import ValidationError as DjangoValidationError
+from django.core.validators import (
+    EmailValidator as _EmailValidator,
+    MaxLengthValidator as _MaxLengthValidator,
+)
+from django.utils.translation import gettext_lazy as _
+
+from glom import assign
+from rest_framework.fields import get_error_detail
+from rest_framework.serializers import ValidationError
+
+from openforms.typing import JSONObject
+
+from .datastructures import FormioConfigurationWrapper, FormioData
+from .typing import Component
+from .utils import get_component_empty_value
+
+missing = object()
+
+
+class RequiredValidator:
+    def __init__(self, empty_value):
+        self.empty_value = empty_value
+
+    def __call__(self, value):
+        if value is missing or value == self.empty_value or value is None:
+            raise ValidationError(
+                _("You must provide a non-empty value."), code="required"
+            )
+        return value
+
+
+class ReturnValueMixin:
+    def __call__(self, value):
+        super().__call__(value)  # type: ignore
+        return value
+
+
+class EmailValidator(ReturnValueMixin, _EmailValidator):
+    pass
+
+
+class MaxLengthValidator(ReturnValueMixin, _MaxLengthValidator):
+    pass
+
+
+Validator = Callable[[Any], Any]
+
+
+class ValidatorChain:
+    def __init__(self, validators: list[Validator]):
+        self.validators = validators
+
+    def __call__(self, value: Any) -> Any:
+        for validator in self.validators:
+            value = validator(value)
+        return value
+
+
+# XXX: if we start using this more broadly, see how we can leverage the component
+# registry.
+def build_validation_chain(component: Component) -> ValidatorChain:
+    assert "type" in component
+    if "validate" not in component:
+        return ValidatorChain([])
+
+    validators = []
+
+    # add required validator, generic for all component types
+    match component["validate"]:
+        case {"required": True}:
+            empty_value = get_component_empty_value(component)
+            validators.append(RequiredValidator(empty_value))
+
+    if component["type"] == "email":
+        validators.append(EmailValidator())
+
+    # text based validators
+    match component:
+        case {
+            "type": "textfield" | "email" | "phoneNumber" | "bsn",
+            "validate": {"maxLength": max_length},
+        } if isinstance(max_length, int):
+            validators.append(MaxLengthValidator(max_length))
+
+    return ValidatorChain(validators)
+
+
+def validate_formio_data(components, values: JSONObject) -> None:
+    """
+    Minimal form.io validation.
+
+    Supports:
+
+    - required
+    - maxLength
+    - email
+
+    """
+    wrapper = FormioConfigurationWrapper({"components": components})
+    paths = wrapper.reverse_flattened
+    data = FormioData(values)
+
+    errors = {}
+
+    for component in wrapper:
+        if "key" not in component:
```

Comment:
Are there components without key?

Reply:
I will turn it into an assert to satisfy the type checker!